### PR TITLE
add complex dtype support to `cupy.std` and `cupy.var`

### PR DESCRIPTION
**Pull request by grlee77**
The variance for a complex-valued array is just the sum of the variances of the real and imaginary parts.

This PR only implements the `out=None` case, but that is probably by far the most common use case.

### CODE REVIEW

File: cupy/core/_routines_statistics.pyx [hunk 1]
```diff
@@ -300,9 +300,15 @@ cdef _nanargmax = create_reduction_func(
 
 cdef ndarray _var(
         ndarray a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
-    assert a.dtype.kind != 'c', 'Variance for complex numbers is not ' \
-                                'implemented. Current implemention does not ' \
-                                'convert the dtype'
+
+    if a.dtype.kind == 'c':
+        assert out is None, 'Variance for complex numbers is not ' \
```
Comment (asi1024) [inline cupy/core/_routines_statistics.pyx]: Could you raise `NotImplementedError`?

Comment (grlee77) [reply cupy/core/_routines_statistics.pyx]: Sure.

**Comment by grlee77** [comment]
I have fixed the cases where the user specifies `dtype='F'` or `dtype='D' `.

> How about fixing _var_core to accept complex type?

I am not sure about how to best do this. I haven't looked too closely at the internals of `ReductionKernel` , but wouldn't that lead to a complex-valued `out` with the variances of the real and imaginary components in the real and imaginary channels? 

In other words if we start with a signal like
```Python
import numpy as np
x = np.arange(5) + 1j * np.arange(5)
```
It seems like the current reduction with complex type `T` would give the following output:
```Python
v = np.var(x.imag) + 1j * np.var(x.imag)
```
-> `(2+2j)`

When what is wanted is either `4` or `(4 + 0j)` depending on the requested output dtype.
(i.e. `np.var(x.imag) + np.var(x.real)`)

If there is a reduction notation that would accomplish that can you point me to a good example that I could follow?

**Comment by asi1024** [comment]
`create_reduction_func` seems useful to accomplish it, but the current implementation also seems OK to temporary implementation for the support. How do you think?

**Comment by grlee77** [comment]
> create_reduction_func seems useful to accomplish it, but the current implementation also seems OK to temporary implementation for the support. How do you think?

I don't think I will have time in the near future to investigate reimplementing this based on `create_reduction_function`.

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
LGTM, but we may create another PR for more performance improvement.
Thank you very much for the PR!

**Pull request merged**